teach/concepts/data-validation/strategy.md

```diff
@@ -0,0 +1,335 @@
+
+#### ValidationError
+
+A `ValidationError` object would likely contain three key pieces of information:
+
+-   `message`: The user-facing error message, translated into the current language.
+-   `dependentProperties`: An array of property object descriptor involved in the validation failure.
+-   `rule`: The validation rule object descriptor that failed (e.g., `isMandatory`, `min`, or a custom rule like `endDateAfterStartDate`).
+
+#### Example 1: Basic Mandatory Field Error
+
+This error occurs when a field marked with `"isMandatory": true` is left empty.
+
+**Scenario:** The user saves an employee record but leaves the `firstName` field blank.
+
+-   **Invalid Data:** `firstName: ""`
+-   **Rule Violated:** The auto-generated `isMandatory` rule on the `firstName` property.
+
+The resulting `ValidationError` would look like this:
+
+```json
+{
+    "message": "First Name is required and cannot be empty.",
+    "dependentProperties": [
+        {
+            "name": "firstName"
+            // ... (all other PropertyDescriptor properties)
+        }
+    ],
+    "rule": {
+        "name": "isMandatory"
+        // ... (all other PropertyDescriptor properties)
+    }
+}
+```
+
+#### Example 2: Custom Single-Property Rule Error
+
+This error is generated by a custom rule that checks a specific format.
+
+**Scenario:** The user enters an `employeeId` that doesn't follow the required pattern.
+
+-   **Invalid Data:** `employeeId: "E-1234"`
+-   **Rule Violated:** The custom `employeeIdFormat` rule.
+
+The resulting `ValidationError` would look like this:
+
+```json
+{
+    "message": "Employee ID must be in the format 'E' followed by 5 digits (e.g., E12345).",
+    "dependentProperties": [{ "name": "employeeId" }],
+    "rule": { "name": "employeeIdFormat" }
+}
+```
+
+#### Example 3: Complex Multi-Property Rule Error
+
+This error involves a rule that compares the values of two different properties.
+
+**Scenario:** The user sets an employee's `endDate` to a date that is before their `startDate`.
+
+-   **Invalid Data:** `startDate: "2025-11-10"`, `endDate: "2025-11-01"`
+-   **Rule Violated:** The custom `endDateAfterStartDate` rule.
+
+The resulting `ValidationError` would look like this:
+
+```json
+{
+    "message": "End date must be after the start date.",
+    "dependentProperties": [{ "name": "startDate" }, { "name": "endDate" }],
+    "rule": { "name": "endDateAfterStartDate" }
+}
+```
+
+#### The Complete `invalidityState` property
+
+When multiple validation rules fail on an object instance, the `invalidityState` property would contain an array of all the corresponding `ValidationError` objects.
+
+For an employee instance with all three errors from the examples above, its `invalidityState` would look like this:
+
+```json
+"invalidityState": [
```

Comment:
I don't think this as just an array works well, as it would force each property field to loop on errors and test their validationProperties to see if it contains their own property. It should be easy and fast for a property field to find errors related to the property (properties if a group). We could have a map with an entry per property and as a value an array  of errors, as one property could be involved on more than one validation rule